File: sysup/internal/tools/tidewave.go
package tools

import (
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"

	"sysup/internal/detect"
	"sysup/internal/download"
)

const (
	tidewaveDefaultPort = "8000"
	tidewaveCLIURL      = "https://github.com/tidewave-ai/tidewave_app/releases/latest/download/tidewave-cli-x86_64-unknown-linux-gnu"
)

func tidewaveAppBinDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "bin")
}

func tidewaveCLIPath() string {
	return filepath.Join(tidewaveAppBinDir(), "tidewave-cli")
}

// RunTidewave ports tidewave/tidewave.sh: install/update/fix-codex-acp
// subcommands, or (with no recognized subcommand) auto-update+fix silently
// and exec the real CLI, injecting a default -p 8000 if none was passed.
func RunTidewave(args []string) error {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	switch sub {
	case "install":
		return tidewaveCmdInstall()
	case "update":
		return tidewaveCmdUpdate(false)
	case "fix-codex-acp":
		return tidewaveCmdFixCodexACP(false)
	case "help", "-h", "--help":
		tidewaveUsage()
		return nil
	default:
		return tidewaveDefaultRun(args)
	}
}

func tidewaveCmdInstall() error {
	dir := tidewaveAppBinDir()
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	sysupPath, err := os.Executable()
	if err != nil {
		sysupPath = "sysup"
	}
	wrapper := fmt.Sprintf("#!/usr/bin/env bash\nexec \"%s\" tidewave \"$@\"\n", sysupPath)
	target := filepath.Join(dir, "tidewave")
	if err := os.WriteFile(target, []byte(wrapper), 0o755); err != nil {
		return err
	}
	fmt.Println("OK: wrapper instalado em", target)
	return nil
}

func tidewaveCmdUpdate(silent bool) error {
	if detect.HasTool("tidewave-update") {
		return runNamed(silent, "tidewave-update")
	}

	dir := tidewaveAppBinDir()
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	tmp, err := os.CreateTemp("", "tidewave-cli-*")
	if err != nil {
		return err
	}
	tmpPath := tmp.Name()
	tmp.Close()
	defer os.Remove(tmpPath)

	if !silent {
		fmt.Println("Baixando:", tidewaveCLIURL)
	}
	if err := download.GetToFile(tidewaveCLIURL, tmpPath); err != nil {
		return err
	}
	if err := os.Chmod(tmpPath, 0o755); err != nil {
		return err
	}
	if err := copyFile(tmpPath, tidewaveCLIPath(), 0o755); err != nil {
		return err
	}
	if !silent {
		fmt.Println("OK:", tidewaveCLIPath(), "atualizado.")
	}
	return nil
}

func tidewaveCmdFixCodexACP(silent bool) error {
	if detect.HasTool("tidewave-fix-codex-acp") {
		return runNamed(silent, "tidewave-fix-codex-acp")
	}

	acp, err := exec.LookPath("codex-acp")
	if err != nil {
		if silent {
			return err
		}
		return fmt.Errorf("codex-acp não encontrado no PATH")
	}

	cacheHome := os.Getenv("XDG_CACHE_HOME")
	if cacheHome == "" {
		home, _ := os.UserHomeDir()
		cacheHome = filepath.Join(home, ".cache")
	}
	downloads := filepath.Join(cacheHome, "tidewave", "downloads")
	if err := os.MkdirAll(downloads, 0o755); err != nil {
		return err
	}

	entries, _ := os.ReadDir(downloads)
	for _, e := range entries {
		if strings.HasPrefix(e.Name(), "codex-acp-linux-x64-") {
			link := filepath.Join(downloads, e.Name())
			os.Remove(link)
			_ = os.Symlink(acp, link)
		}
	}
	pinned := filepath.Join(downloads, "codex-acp-linux-x64-0-8-2")
	os.Remove(pinned)
	_ = os.Symlink(acp, pinned)

	if !silent {
		fmt.Println("OK: codex-acp linkado em", downloads)
	}
	return nil
}

func tidewaveDefaultRun(args []string) error {
	if detect.HasTool("tidewave-update") {
		_ = runNamed(true, "tidewave-update")
	} else {
		_ = tidewaveCmdUpdate(true)
	}

	if detect.HasTool("tidewave-fix-codex-acp") {
		_ = runNamed(true, "tidewave-fix-codex-acp")
	} else {
		_ = tidewaveCmdFixCodexACP(true)
	}

	cliBin := tidewaveCLIPath()
	if stat, err := os.Stat(cliBin); err != nil || stat.Mode()&0o111 == 0 {
		fmt.Println("tidewave-cli não encontrado. Executando update...")
		if err := tidewaveCmdUpdate(false); err != nil {
			return err
		}
	}

	hasPort := false
	for _, a := range args {
		if a == "-p" || a == "--port" {
			hasPort = true
			break
		}
	}
	finalArgs := args
	if !hasPort {
		finalArgs = append([]string{"-p", tidewaveDefaultPort}, args...)
	}

	return syscall.Exec(cliBin, append([]string{cliBin}, finalArgs...), os.Environ())
}

func tidewaveUsage() {
	fmt.Print(`Uso: sysup tidewave [comando] [args]

Comandos:
  install         Instala o wrapper "tidewave" em ~/.local/bin
  update          Atualiza o binário tidewave-cli
  fix-codex-acp   Linka/arruma o codex-acp no cache
  help            Mostra esta ajuda

Sem comando, executa o Tidewave (com update/fix silenciosos)
`)
}

// runNamed runs a named executable with no captured output (or fully
// silenced, when silent is true) — used for delegating to optional
// tidewave-update / tidewave-fix-codex-acp helper scripts if present.
func runNamed(silent bool, name string, args ...string) error {
	cmd := exec.Command(name, args...)
	if !silent {
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
	}
	return cmd.Run()
}

func copyFile(src, dst string, mode os.FileMode) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, mode)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

File: sysup/sysup/src/tools/tidewave.rs
use anyhow::{anyhow, Context};
use std::os::unix::process::CommandExt;
use std::path::PathBuf;
use std::process::{Command, Stdio};

use crate::detect;
use crate::download;

const TIDEWAVE_DEFAULT_PORT: &str = "8000";
const TIDEWAVE_CLI_URL: &str =
    "https://github.com/tidewave-ai/tidewave_app/releases/latest/download/tidewave-cli-x86_64-unknown-linux-gnu";

fn tidewave_app_bin_dir() -> PathBuf {
    let home = std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_default();
    home.join(".local").join("bin")
}

fn tidewave_cli_path() -> PathBuf {
    tidewave_app_bin_dir().join("tidewave-cli")
}

// Ports tidewave/tidewave.sh: install/update/fix-codex-acp subcommands, or
// (with no recognized subcommand) auto-update+fix silently and exec the
// real CLI, injecting a default -p 8000 if none was passed.
pub fn run_tidewave(args: &[String]) -> anyhow::Result<()> {
    let sub = args.first().map(String::as_str).unwrap_or("");
    match sub {
        "install" => tidewave_cmd_install(),
        "update" => tidewave_cmd_update(false),
        "fix-codex-acp" => tidewave_fix_codex_acp(false),
        "help" | "-h" | "--help" => {
            tidewave_usage();
            Ok(())
        }
        _ => tidewave_default_run(args),
    }
}

fn tidewave_cmd_install() -> anyhow::Result<()> {
    let dir = tidewave_app_bin_dir();
    std::fs::create_dir_all(&dir)?;
    let sysup_path = std::env::current_exe()
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "sysup".to_string());
    let wrapper = format!("#!/usr/bin/env bash\nexec \"{sysup_path}\" tidewave \"$@\"\n");
    let target = dir.join("tidewave");
    write_executable(&target, wrapper.as_bytes())?;
    println!("OK: wrapper instalado em {}", target.display());
    Ok(())
}

fn tidewave_cmd_update(silent: bool) -> anyhow::Result<()> {
    if detect::has_tool("tidewave-update") {
        return run_named(silent, "tidewave-update", &[]);
    }

    let dir = tidewave_app_bin_dir();
    std::fs::create_dir_all(&dir)?;

    let tmp_path =
        std::env::temp_dir().join(format!("tidewave-cli-{}", std::process::id()));

    if !silent {
        println!("Baixando: {TIDEWAVE_CLI_URL}");
    }
    download::get_to_file(TIDEWAVE_CLI_URL, &tmp_path)?;

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o755))?;
    }

    let dest = tidewave_cli_path();
    let copy_result = copy_file(&tmp_path, &dest, 0o755);
    let _ = std::fs::remove_file(&tmp_path);
    copy_result?;

    if !silent {
        println!("OK: {} atualizado.", dest.display());
    }
    Ok(())
}

fn tidewave_fix_codex_acp(silent: bool) -> anyhow::Result<()> {
    if detect::has_tool("tidewave-fix-codex-acp") {
        return run_named(silent, "tidewave-fix-codex-acp", &[]);
    }

    let acp = which::which("codex-acp").map_err(|_| {
        if silent {
            anyhow!("codex-acp não encontrado no PATH (silent)")
        } else {
            anyhow!("codex-acp não encontrado no PATH")
        }
    })?;

    let cache_home = std::env::var_os("XDG_CACHE_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            std::env::var_os("HOME")
                .map(PathBuf::from)
                .unwrap_or_default()
                .join(".cache")
        });
    let downloads = cache_home.join("tidewave").join("downloads");
    std::fs::create_dir_all(&downloads)?;

    if let Ok(entries) = std::fs::read_dir(&downloads) {
        for entry in entries.flatten() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            if name.starts_with("codex-acp-linux-x64-") {
                let link = downloads.join(entry.file_name());
                let _ = std::fs::remove_file(&link);
                let _ = std::os::unix::fs::symlink(&acp, &link);
            }
        }
    }
    let pinned = downloads.join("codex-acp-linux-x64-0-8-2");
    let _ = std::fs::remove_file(&pinned);
    let _ = std::os::unix::fs::symlink(&acp, &pinned);

    if !silent {
        println!("OK: codex-acp linkado em {}", downloads.display());
    }
    Ok(())
}

// Uses real process::exec (not spawn) so the tidewave CLI replaces this
// process in place — matching Go's syscall.Exec: no lingering sysup parent,
// signals/tty go straight to the CLI, exit code propagates untouched.
fn tidewave_default_run(args: &[String]) -> anyhow::Result<()> {
    if detect::has_tool("tidewave-update") {
        let _ = run_named(true, "tidewave-update", &[]);
    } else {
        let _ = tidewave_cmd_update(true);
    }

    if detect::has_tool("tidewave-fix-codex-acp") {
        let _ = run_named(true, "tidewave-fix-codex-acp", &[]);
    } else {
        let _ = tidewave_fix_codex_acp(true);
    }

    let cli_bin = tidewave_cli_path();
    let executable = std::fs::metadata(&cli_bin)
        .map(|m| {
            use std::os::unix::fs::PermissionsExt;
            m.permissions().mode() & 0o111 != 0
        })
        .unwrap_or(false);
    if !executable {
        println!("tidewave-cli não encontrado. Executando update...");
        tidewave_cmd_update(false)?;
    }

    let has_port = args.iter().any(|a| a == "-p" || a == "--port");
    let mut final_args: Vec<String> = Vec::new();
    if !has_port {
        final_args.push("-p".to_string());
        final_args.push(TIDEWAVE_DEFAULT_PORT.to_string());
    }
    final_args.extend(args.iter().cloned());

    let err = Command::new(&cli_bin).args(&final_args).exec();
    Err(anyhow!("exec {} falhou: {}", cli_bin.display(), err))
}

fn tidewave_usage() {
    print!(
        r#"Uso: sysup tidewave [comando] [args]

Comandos:
  install         Instala o wrapper "tidewave" em ~/.local/bin
  update          Atualiza o binário tidewave-cli
  fix-codex-acp   Linka/arruma o codex-acp no cache
  help            Mostra esta ajuda

Sem comando, executa o Tidewave (com update/fix silenciosos)
"#
    );
}

// Runs a named executable with no captured output (or fully silenced, when
// silent is true) — used for delegating to optional tidewave-update /
// tidewave-fix-codex-acp helper scripts if present.
fn run_named(silent: bool, name: &str, args: &[String]) -> anyhow::Result<()> {
    let mut cmd = Command::new(name);
    cmd.args(args);
    if silent {
        cmd.stdout(Stdio::null()).stderr(Stdio::null());
    } else {
        cmd.stdout(Stdio::inherit()).stderr(Stdio::inherit());
    }
    let status = cmd
        .status()
        .with_context(|| format!("falha ao executar {name}"))?;
    if !status.success() {
        return Err(anyhow!("{} falhou: {}", name, status));
    }
    Ok(())
}

fn copy_file(src: &std::path::Path, dst: &std::path::Path, mode: u32) -> anyhow::Result<()> {
    use std::os::unix::fs::OpenOptionsExt;

    let mut input = std::fs::File::open(src)?;
    let mut output = std::fs::OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .mode(mode)
        .open(dst)?;
    std::io::copy(&mut input, &mut output)?;
    Ok(())
}

fn write_executable(path: &std::path::Path, content: &[u8]) -> anyhow::Result<()> {
    use std::os::unix::fs::OpenOptionsExt;

    let mut f = std::fs::OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .mode(0o755)
        .open(path)?;
    std::io::Write::write_all(&mut f, content)?;
    Ok(())
}
